# Netflix EDA — Exploratory Data Analysis
**Dataset:** Netflix Movies & TV Shows (Kaggle)
**Tools:** Python, Pandas, NumPy
**Goal:** Uncover trends and patterns in Netflix's content library

# 1. Importing Libraries & Loading Data

In [138]:
# Importing required libraries for data analysis
import numpy as np
import pandas as pd

In [139]:
# Loading the Netflix dataset into a DataFrame
df=pd.read_csv('netflix_titles.csv')

# 2. Basic Inspection

In [140]:
# Checking total rows and columns in the dataset
df.shape

(8807, 12)

In [141]:
# Displaying all column names
df.columns

Index(['show_id', 'type', 'title', 'director', 'cast', 'country', 'date_added',
       'release_year', 'rating', 'duration', 'listed_in', 'description'],
      dtype='object')

In [142]:
# Previewing first 5 rows of the dataset
df.head()

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description
0,s1,Movie,Dick Johnson Is Dead,Kirsten Johnson,NaN,United States,"September 25, 2021",2020,PG-13,90 min,Documentaries,"As her father nears the end of his life, filmm..."
1,s2,TV Show,Blood & Water,NaN,"Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...",South Africa,"September 24, 2021",2021,TV-MA,2 Seasons,"International TV Shows, TV Dramas, TV Mysteries","After crossing paths at a party, a Cape Town t..."
2,s3,TV Show,Ganglands,Julien Leclercq,"Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...",NaN,"September 24, 2021",2021,TV-MA,1 Season,"Crime TV Shows, International TV Shows, TV Act...",To protect his family from a powerful drug lor...
3,s4,TV Show,Jailbirds New Orleans,NaN,NaN,NaN,"September 24, 2021",2021,TV-MA,1 Season,"Docuseries, Reality TV","Feuds, flirtations and toilet talk go down amo..."
4,s5,TV Show,Kota Factory,NaN,"Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...",India,"September 24, 2021",2021,TV-MA,2 Seasons,"International TV Shows, Romantic TV Shows, TV ...",In a city of coaching centers known to train I...


In [143]:
# Getting detailed info about datatypes and null counts
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


In [144]:
# Statistical summary of numerical columns
df.describe()

,release_year
count,8807.000000
mean,2014.180198
std,8.819312
min,1925.000000
25%,2013.000000
50%,2017.000000
75%,2019.000000
max,2021.000000


# 3. Data Cleaning

In [145]:
# Counting missing values in each column
df.isnull().sum()

show_id            0
type               0
title              0
director        2634
cast             825
country          831
date_added        10
release_year       0
rating             4
duration           3
listed_in          0
description        0
dtype: int64

In [146]:
# Dropping rows where director is missing
df.dropna(subset=["director"],inplace=True)

In [147]:
# Filling missing values in country, cast and rating columns
df['country'] = df['country'].fillna('Unknown')
df['cast']=df['cast'].fillna('No Cast')
df['rating']=df['rating'].fillna("No Rating")
df['duration']=df['duration'].fillna('Unknown')

 # 4. DateTime Handling

In [148]:
# Converting date_added to datetime format and extracting year and month
df['date_added'] = pd.to_datetime(df['date_added'], format='mixed')
df['year_added'] = df['date_added'].dt.year
df['month_added'] = df['date_added'].dt.month
df[['date_added', 'year_added', 'month_added']].head()

,date_added,year_added,month_added
0,2021-09-25,2021,9
2,2021-09-24,2021,9
5,2021-09-24,2021,9
6,2021-09-24,2021,9
7,2021-09-24,2021,9


# 5. Content Analysis

In [149]:
# Counting Movies vs TV Shows
df['type'].value_counts()

type
Movie      5943
TV Show     230
Name: count, dtype: int64

In [150]:
# Top 10 countries producing most content
df['country'].value_counts().head(10)

country
United States     2048
India              890
Unknown            422
United Kingdom     212
Canada             119
Spain              106
Egypt               90
Nigeria             90
Japan               85
France              81
Name: count, dtype: int64

In [151]:
# Content added to Netflix per year
df['year_added'].value_counts().sort_index()

year_added
2008       1
2009       2
2010       1
2011      13
2012       3
2013       7
2014      18
2015      57
2016     242
2017     854
2018    1214
2019    1418
2020    1315
2021    1028
Name: count, dtype: int64

In [152]:
# Top 10 directors with most titles on Netflix
df['director'].value_counts().head(10)

director
Rajiv Chilaka             19
Raúl Campos, Jan Suter    18
Marcus Raboy              16
Suhas Kadav               16
Jay Karas                 14
Cathy Garcia-Molina       13
Martin Scorsese           12
Youssef Chahine           12
Jay Chapman               12
Steven Spielberg          11
Name: count, dtype: int64

# 6. String Operations

In [153]:
# Splitting genres and finding top 10 most common genres
df['genre'] = df['listed_in'].str.split(',')
df_genres = df.explode('genre')
df_genres['genre'] = df_genres['genre'].str.strip()

# 7. GroupBy & Aggregation 

In [154]:
# Extracting main genre from listed_in column
df_genres['genre'].value_counts().head(10)
df['main_genre'] = df['listed_in'].str.split(',').str[0].str.strip()
df['main_genre'].value_counts().head(10)

main_genre
Dramas                      1587
Comedies                    1186
Action & Adventure           851
Documentaries                777
Children & Family Movies     570
Stand-Up Comedy              309
Horror Movies                274
International Movies         123
Classic Movies                80
Thrillers                     65
Name: count, dtype: int64

In [155]:
# Content count by country and type for top 10 countries
top_countries = df['country'].value_counts().head(10).index
df[df['country'].isin(top_countries)].groupby(['country','type'])['title'].count()

country         type   
Canada          Movie       116
                TV Show       3
Egypt           Movie        90
France          Movie        75
                TV Show       6
India           Movie       881
                TV Show       9
Japan           Movie        75
                TV Show      10
Nigeria         Movie        86
                TV Show       4
Spain           Movie        95
                TV Show      11
United Kingdom  Movie       191
                TV Show      21
United States   Movie      2000
                TV Show      48
Unknown         Movie       381
                TV Show      41
Name: title, dtype: int64

In [156]:
# Average release year by content type
df.groupby('type')['release_year'].mean()

type
Movie      2013.010601
TV Show    2017.573913
Name: release_year, dtype: float64

In [157]:
# Content added per year broken down by type
df.groupby(['year_added','type'])['title'].count()

year_added  type   
2008        Movie         1
2009        Movie         2
2010        Movie         1
2011        Movie        13
2012        Movie         3
2013        Movie         6
            TV Show       1
2014        Movie        18
2015        Movie        54
            TV Show       3
2016        Movie       234
            TV Show       8
2017        Movie       823
            TV Show      31
2018        Movie      1190
            TV Show      24
2019        Movie      1381
            TV Show      37
2020        Movie      1254
            TV Show      61
2021        Movie       963
            TV Show      65
Name: title, dtype: int64

In [158]:
# Content count by rating category
df.groupby('rating')['title'].count().sort_values(ascending=False)

rating
TV-MA        2117
TV-14        1457
R             795
TV-PG         538
PG-13         489
PG            286
TV-Y7         132
TV-G          118
TV-Y          112
NR             75
G              41
TV-Y7-FV        4
UR              3
NC-17           2
74 min          1
No Rating       1
84 min          1
66 min          1
Name: title, dtype: int64

# 8. Pivot Tables

In [159]:
# Pivot table showing content count by type and rating
pivot = pd.pivot_table(df, values='title', 
                       index='rating', 
                       columns='type', 
                       aggfunc='count')
pivot

type,Movie,TV Show
rating,,
66 min,1.0,NaN
74 min,1.0,NaN
84 min,1.0,NaN
G,41.0,NaN
NC-17,2.0,NaN
NR,75.0,NaN
No Rating,1.0,NaN
PG,286.0,NaN
PG-13,489.0,NaN


In [160]:
# Pivot table showing content added per year by type
pivot2 = pd.pivot_table(df, values='title',
                        index='year_added',
                        columns='type',
                        aggfunc='count')
pivot2

type,Movie,TV Show
year_added,,
2008,1.0,NaN
2009,2.0,NaN
2010,1.0,NaN
2011,13.0,NaN
2012,3.0,NaN
2013,6.0,1.0
2014,18.0,NaN
2015,54.0,3.0
2016,234.0,8.0


# 9. Interesting Insights

In [161]:
# Which month has the most content added on Netflix
df['month_added'].value_counts().sort_index()

month_added
1     537
2     385
3     542
4     549
5     446
6     495
7     577
8     530
9     520
10    546
11    497
12    549
Name: count, dtype: int64

In [162]:
# Total number of unique directors on Netflix
df['director'].nunique()

4528

In [163]:
# Top 10 most featured cast members on Netflix
df['cast'].str.split(',').explode().str.strip().value_counts().head(10)

cast
No Cast             473
Anupam Kher          42
Shah Rukh Khan       35
Naseeruddin Shah     32
Akshay Kumar         30
Om Puri              30
Paresh Rawal         28
Julie Tejwani        28
Amitabh Bachchan     28
Boman Irani          27
Name: count, dtype: int64

In [164]:
# Finding the 5 longest movies on Netflix
movies = df[df['type']=='Movie'].copy()
movies = movies[movies['duration'] != 'Unknown']
movies['duration_mins'] = movies['duration'].str.replace(' min','').astype(int)
movies.nlargest(5, 'duration_mins')[['title','duration_mins']]

,title,duration_mins
2491,The School of Mischief,253
2487,No Longer kids,237
2484,Lock Your Girls In,233
2488,Raya and Sakina,230
166,Once Upon a Time in America,229


## Key Findings
- United States dominates Netflix with the most content
- 2019 was the peak year for content added on Netflix
- International Movies is the most common genre
- Anupam Kher is the most featured actor on Netflix
- The longest movie on Netflix is The School of Mischief at 253 mins